File: packages/sdk-ts/src/prepaid-wallets.ts
import type { OAuthAgentClient } from './oauth-agent.js';

export type PrepaidCustodyMode = 'sandbox_ledger' | 'external';
export type PrepaidWalletStatus = 'active' | 'blocked' | 'closed';
export type WalletAssignmentStatus = 'active' | 'blocked' | 'revoked';
export type WalletSpendPolicyScope = 'assignment' | 'wallet' | 'agent' | 'group' | 'principal' | 'developer';
export type WalletSpendPolicyEffect = 'limit' | 'deny' | 'require_approval';
export type WalletSpendPolicyWindow = 'per_authorization' | 'rolling' | 'calendar_day' | 'calendar_week' | 'calendar_month' | 'lifetime';
export type WalletSpendPolicyStatus = 'active' | 'disabled' | 'revoked';

export interface PrepaidWallet {
  walletId: string;
  principalId: string;
  name: string;
  custodyMode: PrepaidCustodyMode;
  provider: string | null;
  providerWalletId: string | null;
  walletAddress: string | null;
  network: string;
  asset: string;
  decimals: number;
  availableAmount: string;
  reservedAmount: string;
  lowBalanceThreshold: string;
  maxBalance: string | null;
  maxReloadAmount: string | null;
  reloadCumulativeLimit: string | null;
  reloadPeriodSeconds: number | null;
  reloadCountLimit: number | null;
  status: PrepaidWalletStatus;
  blockedAt: string | null;
  blockedReason: string | null;
  metadata: Record<string, unknown>;
  createdAt: string;
  updatedAt: string;
}

/** Agent-visible wallet projection. Custody provider identifiers and metadata are never exposed. */
export interface AssignedPrepaidWallet {
  walletId: string;
  name: string;
  network: string;
  asset: string;
  decimals: number;
  availableAmount: string;
  reservedAmount: string;
  lowBalanceThreshold: string;
  status: PrepaidWalletStatus;
  blockedAt: string | null;
  blockedReason: string | null;
  assignmentId: string;
  assignmentStatus: WalletAssignmentStatus;
  perTransactionLimit: string;
  cumulativeLimit: string;
  cumulativePeriodSeconds: number;
  allowedRecipients: string[];
  allowedScopes: string[];
  allowedResourceOrigins: string[];
  allowAnyRecipient: boolean;
  allowAnyScope: boolean;
  allowAnyResource: boolean;
  budgetGroup: string | null;
  validFrom: string;
  validUntil: string | null;
  allWalletsBlocked: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface WalletAssignment {
  assignmentId: string;
  walletId: string;
  agentId: string;
  principalId: string;
  status: WalletAssignmentStatus;
  perTransactionLimit: string;
  cumulativeLimit: string;
  cumulativePeriodSeconds: number;
  allowedRecipients: string[];
  allowedScopes: string[];
  allowedResourceOrigins: string[];
  allowAnyRecipient: boolean;
  allowAnyScope: boolean;
  allowAnyResource: boolean;
  budgetGroup: string | null;
  validFrom: string;
  validUntil: string | null;
  blockedAt: string | null;
  blockedReason: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface CreatePrepaidWalletParams {
  name: string;
  custodyMode: PrepaidCustodyMode;
  provider?: string;
  providerWalletId?: string;
  walletAddress?: string;
  network: string;
  asset: string;
  decimals?: number;
  lowBalanceThreshold?: string;
  maxBalance?: string;
  maxReloadAmount?: string;
  reloadCumulativeLimit?: string;
  reloadPeriodSeconds?: number;
  reloadCountLimit?: number;
  metadata?: Record<string, unknown>;
}

export interface AssignPrepaidWalletParams {
  agentId: string;
  perTransactionLimit: string;
  cumulativeLimit: string;
  cumulativePeriodSeconds: number;
  allowedRecipients?: string[];
  allowedScopes?: string[];
  allowedResourceOrigins?: string[];
  allowAnyRecipient?: boolean;
  allowAnyScope?: boolean;
  allowAnyResource?: boolean;
  budgetGroup?: string;
  validUntil?: string;
}

export interface PrepaidAuthorizationRequest {
  walletId?: string;
  amount: string;
  asset: string;
  network: string;
  recipient: string;
  resource: string;
  scope: string;
  maxTimeoutSeconds: number;
  idempotencyKey: string;
  approvalRequestId?: string;
  merchantId?: string;
  purpose?: string;
  projectId?: string;
  costCenter?: string;
}

export interface PrepaidAuthorization {
  authorization: string;
  reservationId: string;
  walletId: string;
  assignmentId: string;
  amount: string;
  asset: string;
  network: string;
  recipient: string;
  expiresAt: string;
  remainingAvailable: string;
  remainingCumulative: string;
  policyDecisionId: string | null;
}

export interface PrepaidApprovalRequired {
  status: 'approval_required';
  approvalRequestId: string;
  walletId: string;
  assignmentId: string;
  policyIds: string[];
  expiresAt: string;
}

export type PrepaidAuthorizationResponse = PrepaidAuthorization | PrepaidApprovalRequired;

export interface WalletSpendPolicyInput {
  name: string;
  description?: string;
  scopeType: WalletSpendPolicyScope;
  scopeId?: string;
  effect: WalletSpendPolicyEffect;
  onExceed?: 'deny' | 'require_approval';
  maxAmount?: string;
  maxCount?: number;
  windowType?: WalletSpendPolicyWindow;
  windowSeconds?: number;
  recipients?: string[];
  resourceOrigins?: string[];
  actionScopes?: string[];
  assets?: string[];
  networks?: string[];
  merchantIds?: string[];
  purposes?: string[];
  projectIds?: string[];
  costCenters?: string[];
  requireVerifiedMerchant?: boolean;
  priority?: number;
  validFrom?: string;
  validUntil?: string;
}

export interface WalletSpendPolicy {
  policyId: string;
  developerId: string;
  principalId: string | null;
  name: string;
  description: string | null;
  scopeType: WalletSpendPolicyScope;
  scopeId: string | null;
  effect: WalletSpendPolicyEffect;
  onExceed: 'deny' | 'require_approval';
  maxAmount: string | null;
  maxCount: number | null;
  windowType: WalletSpendPolicyWindow;
  windowSeconds: number | null;
  recipients: string[];
  resourceOrigins: string[];
  actionScopes: string[];
  assets: string[];
  networks: string[];
  merchantIds: string[];
  purposes: string[];
  projectIds: string[];
  costCenters: string[];
  requireVerifiedMerchant: boolean;
  priority: number;
  status: WalletSpendPolicyStatus;
  version: number;
  validFrom: string;
  validUntil: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface WalletPaymentApproval {
  approvalRequestId: string;
  walletId: string;
  assignmentId: string;
  agentId: string;
  amount: string;
  asset: string;
  network: string;
  recipient: string;
  resource: string;
  scope: string;
  merchantId: string | null;
  purpose: string | null;
  projectId: string | null;
  costCenter: string | null;
  policyIds: string[];
  status: 'pending' | 'approved' | 'rejected' | 'consumed' | 'expired';
  reason: string | null;
  expiresAt: string;
  decidedAt: string | null;
  consumedAt: string | null;
  reservationId: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface WalletReloadRequest {
  reloadRequestId: string;
  walletId: string;
  assignmentId: string | null;
  agentId: string | null;
  amount: string;
  reason: string | null;
  status: 'pending' | 'approved' | 'rejected' | 'cancelled' | 'funded';
  requestedBy: 'agent' | 'principal';
  externalReference: string | null;
  createdAt: string;
  decidedAt: string | null;
  fundedAt: string | null;
}

export interface PrepaidWalletActivity {
  assignments: WalletAssignment[];
  reservations: Array<Record<string, unknown>>;
  reloadRequests: WalletReloadRequest[];
  ledger: Array<Record<string, unknown>>;
  policyDecisions: Array<Record<string, unknown>>;
  paymentApprovals: WalletPaymentApproval[];
}

interface DpopResourceClient {
  fetch(input: string | URL, accessToken: string, init?: RequestInit): Promise<Response>;
}

export class PrepaidWalletAgentClient {
  readonly #oauthClient: DpopResourceClient;
  readonly #resourceUrl: string;
  #accessToken: string;

  constructor(options: {
    oauthClient: OAuthAgentClient | DpopResourceClient;
    accessToken: string;
    resourceUrl?: string;
  }) {
    if (!options.accessToken) throw new Error('accessToken is required');
    this.#oauthClient = options.oauthClient;
    const inferred = 'resource' in options.oauthClient
      ? String(options.oauthClient.resource)
      : undefined;
    this.#resourceUrl = normalizeUrl(options.resourceUrl ?? inferred ?? '');
    if (!this.#resourceUrl.endsWith('/v1/prepaid-wallets')) {
      throw new Error('resourceUrl must end with /v1/prepaid-wallets and match the OAuth resource audience');
    }
    this.#accessToken = options.accessToken;
  }

  setAccessToken(accessToken: string): void {
    if (!accessToken) throw new Error('accessToken is required');
    this.#accessToken = accessToken;
  }

  async list(): Promise<AssignedPrepaidWallet[]> {
    const response = await this.#oauthClient.fetch(this.#resourceUrl, this.#accessToken, {
      headers: { Accept: 'application/json' },
    });
    return (await walletJson<{ wallets: AssignedPrepaidWallet[] }>(response)).wallets;
  }

  async authorizePayment(params: PrepaidAuthorizationRequest): Promise<PrepaidAuthorizationResponse> {
    const response = await this.#oauthClient.fetch(
      `${this.#resourceUrl}/authorizations`,
      this.#accessToken,
      jsonRequest(params),
    );
    return walletJson<PrepaidAuthorizationResponse>(response);
  }

  async requestReload(
    walletId: string,
    amount: string,
    idempotencyKey: string,
    reason?: string,
  ): Promise<WalletReloadRequest> {
    const response = await this.#oauthClient.fetch(
      `${this.#resourceUrl}/${encodeURIComponent(walletId)}/reload-requests`,
      this.#accessToken,
      jsonRequest({ amount, idempotencyKey, ...(reason !== undefined ? { reason } : {}) }),
    );
    return walletJson<WalletReloadRequest>(response);
  }

  /** Callback passed directly to @grantex/x402 createX402Agent(). */
  readonly x402Authorizer = (params: PrepaidAuthorizationRequest) => this.authorizePayment(params);
}

export class PrincipalPrepaidWalletClient {
  readonly #baseUrl: string;
  #sessionToken: string;

  constructor(options: { baseUrl: string; sessionToken: string }) {
    this.#baseUrl = normalizeUrl(options.baseUrl);
    if (!options.sessionToken) throw new Error('sessionToken is required');
    this.#sessionToken = options.sessionToken;
  }

  setSessionToken(sessionToken: string): void {
    if (!sessionToken) throw new Error('sessionToken is required');
    this.#sessionToken = sessionToken;
  }

  create(params: CreatePrepaidWalletParams): Promise<PrepaidWallet> {
    return this.#request('/v1/principal/prepaid-wallets', jsonRequest(params));
  }

  async list(): Promise<PrepaidWallet[]> {
    return (await this.#request<{ wallets: PrepaidWallet[] }>('/v1/principal/prepaid-wallets')).wallets;
  }

  activity(walletId: string): Promise<PrepaidWalletActivity> {
    return this.#request(`/v1/principal/prepaid-wallets/${encodeURIComponent(walletId)}/activity`);
  }

  assign(walletId: string, params: AssignPrepaidWalletParams): Promise<WalletAssignment> {
    return this.#request(
      `/v1/principal/prepaid-wallets/${encodeURIComponent(walletId)}/assignments`,
      jsonRequest(params),
    );
  }

  setAssignmentStatus(assignmentId: string, status: WalletAssignmentStatus, reason?: string): Promise<WalletAssignment> {
    return this.#request(
      `/v1/principal/prepaid-wallet-assignments/${encodeURIComponent(assignmentId)}`,
      jsonRequest({ status, ...(reason !== undefined ? { reason } : {}) }, 'PATCH'),
    );
  }

  setWalletStatus(walletId: string, status: PrepaidWalletStatus, reason?: string): Promise<PrepaidWallet> {
    return this.#request(
      `/v1/principal/prepaid-wallets/${encodeURIComponent(walletId)}/status`,
      jsonRequest({ status, ...(reason !== undefined ? { reason } : {}) }, 'PATCH'),
    );
  }

  setAgentBlocked(agentId: string, blocked: boolean, reason?: string): Promise<{ agentId: string; allWalletsBlocked: boolean; reason: string | null }> {
    return this.#request(
      `/v1/principal/prepaid-wallet-agents/${encodeURIComponent(agentId)}/block`,
      jsonRequest({ blocked, ...(reason !== undefined ? { reason } : {}) }, 'PUT'),
    );
  }

  reload(walletId: string, amount: string, idempotencyKey: string, externalReference?: string): Promise<WalletReloadRequest> {
    return this.#request(
      `/v1/principal/prepaid-wallets/${encodeURIComponent(walletId)}/reloads`,
      jsonRequest({ amount, idempotencyKey, ...(externalReference !== undefined ? { externalReference } : {}) }),
    );
  }

  decideReload(requestId: string, decision: 'approved' | 'rejected'): Promise<WalletReloadRequest> {
    return this.#request(
      `/v1/principal/prepaid-wallet-reload-requests/${encodeURIComponent(requestId)}/decision`,
      jsonRequest({ decision }),
    );
  }

  fundReload(requestId: string, externalReference?: string): Promise<WalletReloadRequest> {
    return this.#request(
      `/v1/principal/prepaid-wallet-reload-requests/${encodeURIComponent(requestId)}/fund`,
      jsonRequest({ ...(externalReference !== undefined ? { externalReference } : {}) }),
    );
  }

  releaseReservation(reservationId: string, reason: string): Promise<{ reservationId: string; status: 'released' }> {
    return this.#request(
      `/v1/principal/prepaid-wallet-reservations/${encodeURIComponent(reservationId)}/release`,
      jsonRequest({ reason }),
    );
  }

  createSpendPolicy(params: WalletSpendPolicyInput): Promise<WalletSpendPolicy> {
    return this.#request('/v1/principal/prepaid-wallet-spend-policies', jsonRequest(params));
  }

  async listSpendPolicies(): Promise<WalletSpendPolicy[]> {
    return (await this.#request<{ policies: WalletSpendPolicy[] }>('/v1/principal/prepaid-wallet-spend-policies')).policies;
  }

  setSpendPolicyStatus(policyId: string, status: WalletSpendPolicyStatus): Promise<WalletSpendPolicy> {
    return this.#request(
      `/v1/principal/prepaid-wallet-spend-policies/${encodeURIComponent(policyId)}/status`,
      jsonRequest({ status }, 'PATCH'),
    );
  }

  async listPaymentApprovals(): Promise<WalletPaymentApproval[]> {
    return (await this.#request<{ approvals: WalletPaymentApproval[] }>('/v1/principal/prepaid-wallet-payment-approvals')).approvals;
  }

  decidePaymentApproval(
    approvalRequestId: string,
    decision: 'approved' | 'rejected',
    reason?: string,
  ): Promise<WalletPaymentApproval> {
    return this.#request(
      `/v1/principal/prepaid-wallet-payment-approvals/${encodeURIComponent(approvalRequestId)}/decision`,
      jsonRequest({ decision, ...(reason !== undefined ? { reason } : {}) }),
    );
  }

  async #request<T>(path: string, init: RequestInit = {}): Promise<T> {
    const headers = new Headers(init.headers);
    headers.set('Authorization', `Bearer ${this.#sessionToken}`);
    headers.set('Accept', 'application/json');
    const response = await fetch(`${this.#baseUrl}${path}`, { ...init, headers });
    return walletJson<T>(response);
  }
}

/** Developer API-key client for organization-wide and cross-principal wallet policy. */
export class DeveloperPrepaidWalletPolicyClient {
  readonly #baseUrl: string;
  #apiKey: string;

  constructor(options: { baseUrl: string; apiKey: string }) {
    this.#baseUrl = normalizeUrl(options.baseUrl);
    if (!options.apiKey) throw new Error('apiKey is required');
    this.#apiKey = options.apiKey;
  }

  setApiKey(apiKey: string): void {
    if (!apiKey) throw new Error('apiKey is required');
    this.#apiKey = apiKey;
  }

  create(params: WalletSpendPolicyInput): Promise<WalletSpendPolicy> {
    return this.#request('/v1/prepaid-wallet-spend-policies', jsonRequest(params));
  }

  async list(): Promise<WalletSpendPolicy[]> {
    return (await this.#request<{ policies: WalletSpendPolicy[] }>('/v1/prepaid-wallet-spend-policies')).policies;
  }

  setStatus(policyId: string, status: WalletSpendPolicyStatus): Promise<WalletSpendPolicy> {
    return this.#request(
      `/v1/prepaid-wallet-spend-policies/${encodeURIComponent(policyId)}/status`,
      jsonRequest({ status }, 'PATCH'),
    );
  }

  async #request<T>(path: string, init: RequestInit = {}): Promise<T> {
    const headers = new Headers(init.headers);
    headers.set('Authorization', `Bearer ${this.#apiKey}`);
    headers.set('Accept', 'application/json');
    return walletJson<T>(await fetch(`${this.#baseUrl}${path}`, { ...init, headers }));
  }
}

function normalizeUrl(value: string): string {
  if (!value) throw new Error('A base URL is required');
  const url = new URL(value);
  if (!['http:', 'https:'].includes(url.protocol)) {
    throw new Error('Base URLs must use HTTP or HTTPS');
  }
  if (url.protocol === 'http:' && !isLoopback(url.hostname)) {
    throw new Error('Remote wallet endpoints must use HTTPS');
  }
  if (url.username || url.password) throw new Error('Base URLs must not contain credentials');
  if (url.search || url.hash) throw new Error('Base URLs must not contain a query or fragment');
  return url.toString().replace(/\/$/, '');
}

function isLoopback(hostname: string): boolean {
  const normalized = hostname.replace(/^\[|\]$/g, '').toLowerCase();
  return normalized === 'localhost' || normalized === '127.0.0.1' || normalized === '::1';
}

function jsonRequest(body: unknown, method = 'POST'): RequestInit {
  return {
    method,
    headers: { 'Content-Type': 'application/json', Accept: 'application/json' },
    body: JSON.stringify(body),
  };
}

async function walletJson<T>(response: Response): Promise<T> {
  const body = await response.json().catch(() => undefined) as unknown;
  if (!response.ok) {
    const object = body && typeof body === 'object' ? body as Record<string, unknown> : {};
    const error = new Error(typeof object['message'] === 'string'
      ? object['message']
      : `Prepaid wallet request failed with HTTP ${response.status}`);
    Object.assign(error, {
      name: 'PrepaidWalletApiError',
      status: response.status,
      code: object['code'],
      requestId: object['requestId'],
    });
    throw error;
  }
  return body as T;
}

File: packages/sdk-py/src/grantex/prepaid_wallets.py
from __future__ import annotations

import base64
import hashlib
import ipaddress
import time
import uuid
from typing import Any, List
from urllib.parse import quote, urlsplit, urlunsplit

import httpx
import jwt
from cryptography.hazmat.primitives.asymmetric import ec

from ._errors import GrantexApiError, GrantexAuthError
from ._http import HttpClient

WalletRecord = dict[str, Any]


class WalletSpendPoliciesClient:
    """Developer API-key access to organization-wide wallet spend policy."""

    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def create(self, params: WalletRecord) -> WalletRecord:
        return _record(self._http.post("/v1/prepaid-wallet-spend-policies", params))

    def list(self) -> list[WalletRecord]:
        data = _record(self._http.get("/v1/prepaid-wallet-spend-policies"))
        return _records(data.get("policies"))

    def set_status(self, policy_id: str, status: str) -> WalletRecord:
        return _record(
            self._http.patch(
                f"/v1/prepaid-wallet-spend-policies/{quote(policy_id, safe='')}/status",
                {"status": status},
            )
        )


class PrincipalPrepaidWalletClient:
    """Principal-session wallet, assignment, reload, policy, and approval client."""

    def __init__(
        self,
        *,
        base_url: str,
        session_token: str,
        timeout: float = 30.0,
        client: httpx.Client | None = None,
    ) -> None:
        if not session_token.strip():
            raise ValueError("session_token is required")
        self._base_url = _base_url(base_url)
        self._session_token = session_token.strip()
        self._client = client or httpx.Client(timeout=timeout)
        self._owns_client = client is None

    def set_session_token(self, session_token: str) -> None:
        if not session_token.strip():
            raise ValueError("session_token is required")
        self._session_token = session_token.strip()

    def create(self, params: WalletRecord) -> WalletRecord:
        return self._request("POST", "/v1/principal/prepaid-wallets", params)

    def list(self) -> list[WalletRecord]:
        return _records(
            self._request("GET", "/v1/principal/prepaid-wallets").get("wallets")
        )

    def activity(self, wallet_id: str) -> WalletRecord:
        return self._request(
            "GET", f"/v1/principal/prepaid-wallets/{quote(wallet_id, safe='')}/activity"
        )

    def assign(self, wallet_id: str, params: WalletRecord) -> WalletRecord:
        return self._request(
            "POST",
            f"/v1/principal/prepaid-wallets/{quote(wallet_id, safe='')}/assignments",
            params,
        )

    def set_assignment_status(
        self, assignment_id: str, status: str, reason: str | None = None
    ) -> WalletRecord:
        return self._request(
            "PATCH",
            f"/v1/principal/prepaid-wallet-assignments/{quote(assignment_id, safe='')}",
            _with_reason(status=status, reason=reason),
        )

    def set_wallet_status(
        self, wallet_id: str, status: str, reason: str | None = None
    ) -> WalletRecord:
        return self._request(
            "PATCH",
            f"/v1/principal/prepaid-wallets/{quote(wallet_id, safe='')}/status",
            _with_reason(status=status, reason=reason),
        )

    def set_agent_blocked(
        self, agent_id: str, blocked: bool, reason: str | None = None
    ) -> WalletRecord:
        body: WalletRecord = {"blocked": blocked}
        if reason is not None:
            body["reason"] = reason
        return self._request(
            "PUT",
            f"/v1/principal/prepaid-wallet-agents/{quote(agent_id, safe='')}/block",
            body,
        )

    def reload(
        self,
        wallet_id: str,
        amount: str,
        idempotency_key: str,
        external_reference: str | None = None,
    ) -> WalletRecord:
        body: WalletRecord = {"amount": amount, "idempotencyKey": idempotency_key}
        if external_reference is not None:
            body["externalReference"] = external_reference
        return self._request(
            "POST",
            f"/v1/principal/prepaid-wallets/{quote(wallet_id, safe='')}/reloads",
            body,
        )

    def decide_reload(self, request_id: str, decision: str) -> WalletRecord:
        return self._request(
            "POST",
            f"/v1/principal/prepaid-wallet-reload-requests/{quote(request_id, safe='')}/decision",
            {"decision": decision},
        )

    def fund_reload(
        self, request_id: str, external_reference: str | None = None
    ) -> WalletRecord:
        body: WalletRecord = {}
        if external_reference is not None:
            body["externalReference"] = external_reference
        return self._request(
            "POST",
            f"/v1/principal/prepaid-wallet-reload-requests/{quote(request_id, safe='')}/fund",
            body,
        )

    def release_reservation(self, reservation_id: str, reason: str) -> WalletRecord:
        return self._request(
            "POST",
            f"/v1/principal/prepaid-wallet-reservations/{quote(reservation_id, safe='')}/release",
            {"reason": reason},
        )

    def create_spend_policy(self, params: WalletRecord) -> WalletRecord:
        return self._request(
            "POST", "/v1/principal/prepaid-wallet-spend-policies", params
        )

    def list_spend_policies(self) -> List[WalletRecord]:
        return _records(
            self._request("GET", "/v1/principal/prepaid-wallet-spend-policies").get(
                "policies"
            )
        )

    def set_spend_policy_status(self, policy_id: str, status: str) -> WalletRecord:
        return self._request(
            "PATCH",
            f"/v1/principal/prepaid-wallet-spend-policies/{quote(policy_id, safe='')}/status",
            {"status": status},
        )

    def list_payment_approvals(self) -> List[WalletRecord]:
        return _records(
            self._request("GET", "/v1/principal/prepaid-wallet-payment-approvals").get(
                "approvals"
            )
        )

    def decide_payment_approval(
        self, approval_request_id: str, decision: str, reason: str | None = None
    ) -> WalletRecord:
        return self._request(
            "POST",
            f"/v1/principal/prepaid-wallet-payment-approvals/{quote(approval_request_id, safe='')}/decision",
            _with_reason(decision=decision, reason=reason),
        )

    def _request(
        self, method: str, path: str, body: WalletRecord | None = None
    ) -> WalletRecord:
        response = self._client.request(
            method,
            f"{self._base_url}{path}",
            headers={
                "Authorization": f"Bearer {self._session_token}",
                "Accept": "application/json",
            },
            json=body,
        )
        return _wallet_response(response)

    def close(self) -> None:
        if self._owns_client:
            self._client.close()

    def __enter__(self) -> PrincipalPrepaidWalletClient:
        return self

    def __exit__(self, *args: object) -> None:
        self.close()


class AgentPrepaidWalletClient:
    """DPoP-bound agent wallet client using an ES256 OAuth token key."""

    def __init__(
        self,
        *,
        access_token: str,
        private_key: ec.EllipticCurvePrivateKey,
        resource_url: str,
        timeout: float = 30.0,
        client: httpx.Client | None = None,
    ) -> None:
        if not access_token:
            raise ValueError("access_token is required")
        resource = _base_url(resource_url)
        if not resource.endswith("/v1/prepaid-wallets"):
            raise ValueError("resource_url must end with /v1/prepaid-wallets")
        if not isinstance(private_key.curve, ec.SECP256R1):
            raise ValueError("private_key must use P-256 for ES256 DPoP")
        self._access_token = access_token
        self._private_key = private_key
        self._resource_url = resource
        self._client = client or httpx.Client(timeout=timeout)
        self._owns_client = client is None

    def set_access_token(self, access_token: str) -> None:
        if not access_token:
            raise ValueError("access_token is required")
        self._access_token = access_token

    def list(self) -> list[WalletRecord]:
        return _records(self._request("GET", self._resource_url).get("wallets"))

    def authorize_payment(self, params: WalletRecord) -> WalletRecord:
        return self._request("POST", f"{self._resource_url}/authorizations", params)

    def request_reload(
        self,
        wallet_id: str,
        amount: str,
        idempotency_key: str,
        reason: str | None = None,
    ) -> WalletRecord:
        body: WalletRecord = {"amount": amount, "idempotencyKey": idempotency_key}
        if reason is not None:
            body["reason"] = reason
        return self._request(
            "POST",
            f"{self._resource_url}/{quote(wallet_id, safe='')}/reload-requests",
            body,
        )

    def _request(
        self, method: str, url: str, body: WalletRecord | None = None
    ) -> WalletRecord:
        proof = _dpop_proof(method, url, self._access_token, self._private_key)
        response = self._client.request(
            method,
            url,
            headers={
                "Authorization": f"DPoP {self._access_token}",
                "DPoP": proof,
                "Accept": "application/json",
            },
            json=body,
        )
        return _wallet_response(response)

    def close(self) -> None:
        if self._owns_client:
            self._client.close()

    def __enter__(self) -> AgentPrepaidWalletClient:
        return self

    def __exit__(self, *args: object) -> None:
        self.close()


def generate_dpop_key() -> ec.EllipticCurvePrivateKey:
    return ec.generate_private_key(ec.SECP256R1())


def _dpop_proof(
    method: str, url: str, access_token: str, key: ec.EllipticCurvePrivateKey
) -> str:
    public = key.public_key().public_numbers()
    jwk = {
        "kty": "EC",
        "crv": "P-256",
        "x": _b64(public.x.to_bytes(32, "big")),
        "y": _b64(public.y.to_bytes(32, "big")),
    }
    claims = {
        "htm": method.upper(),
        "htu": url,
        "iat": int(time.time()),
        "jti": str(uuid.uuid4()),
        "ath": _b64(hashlib.sha256(access_token.encode("utf-8")).digest()),
    }
    return str(
        jwt.encode(
            claims, key, algorithm="ES256", headers={"typ": "dpop+jwt", "jwk": jwk}
        )
    )


def _wallet_response(response: httpx.Response) -> WalletRecord:
    try:
        body: Any = response.json()
    except Exception:
        body = None
    if not response.is_success:
        message = (
            body.get("message")
            if isinstance(body, dict)
            else f"HTTP {response.status_code}"
        )
        code = body.get("code") if isinstance(body, dict) else None
        request_id = (
            body.get("requestId")
            if isinstance(body, dict)
            else response.headers.get("x-request-id")
        )
        error = GrantexApiError(
            str(message), response.status_code, body, request_id, code
        )
        if response.status_code in (401, 403):
            raise GrantexAuthError(
                str(message), response.status_code, body, request_id, code
            )
        raise error
    return _record(body)


def _record(value: Any) -> WalletRecord:
    if not isinstance(value, dict):
        raise ValueError("Grantex wallet API returned a non-object response")
    return {str(key): item for key, item in value.items()}


def _records(value: Any) -> list[WalletRecord]:
    if not isinstance(value, list):
        raise ValueError("Grantex wallet API returned a non-list collection")
    return [_record(item) for item in value]


def _with_reason(*, reason: str | None, **values: Any) -> WalletRecord:
    result: WalletRecord = dict(values)
    if reason is not None:
        result["reason"] = reason
    return result


def _base_url(value: str) -> str:
    parsed = urlsplit(value)
    if (
        parsed.scheme not in ("http", "https")
        or not parsed.netloc
        or parsed.query
        or parsed.fragment
    ):
        raise ValueError(
            "base URL must be an absolute HTTP(S) URL without query or fragment"
        )
    if parsed.username or parsed.password:
        raise ValueError("base URL must not contain credentials")
    host = parsed.hostname or ""
    try:
        loopback = ipaddress.ip_address(host).is_loopback
    except ValueError:
        loopback = host.lower() == "localhost"
    if parsed.scheme == "http" and not loopback:
        raise ValueError("remote wallet endpoints must use HTTPS")
    return urlunsplit((parsed.scheme, parsed.netloc, parsed.path.rstrip("/"), "", ""))


def _b64(value: bytes) -> str:
    return base64.urlsafe_b64encode(value).rstrip(b"=").decode("ascii")
